File: tests/fixtures/zig/variables.zig
const std = @import("std");

pub var zig_global: i64 = 73;
pub var zig_sink: i64 = 0;
pub const zig_constant: i64 = 37;
var input_flag: bool = true;
var input_signed: i32 = -42;
var input_unsigned: u64 = 42;
var input_single: f32 = 1.25;
var input_double: f64 = -2.5;
var pointer_parameter_value: i32 = 42;

const SignedAlias = i32;
const UnsignedAlias = u64;
const PointerAlias = i32;
const PointerPair = struct {
    first: i32,
    second: i32,
};
const PointerNode = struct {
    next: ?*const @This(),
    value: i32,
};

noinline fn inspectScalars(
    flag: bool,
    signed_value: SignedAlias,
    unsigned_value: UnsignedAlias,
    single: f32,
    double_precision: f64,
) bool {
    const local_flag = !flag;
    const local_signed: SignedAlias = signed_value + 1;
    const local_unsigned: UnsignedAlias = unsigned_value + 2;
    const local_single: f32 = single + 0.5;
    const local_double: f64 = double_precision - 0.25;
    zig_sink = local_signed;
    std.mem.doNotOptimizeAway(&local_flag);
    std.mem.doNotOptimizeAway(&local_unsigned);
    std.mem.doNotOptimizeAway(&local_single);
    std.mem.doNotOptimizeAway(&local_double);
    return !local_flag and local_signed == -41 and local_unsigned == 44 and
        local_single == 1.75 and local_double == -2.75;
}

noinline fn inspectScopes(value: i32) i32 {
    const outer_value = value + 1;
    {
        const nested_value = outer_value + 1;
        zig_sink = nested_value;
        std.mem.doNotOptimizeAway(&nested_value);
    }
    std.mem.doNotOptimizeAway(&outer_value);
    return outer_value;
}

noinline fn inspectPointers(value: i32, pointer_parameter: *const i32) bool {
    var pointee = value + 2;
    const pointer: *i32 = &pointee;
    const const_pointer: *const i32 = &pointee;
    const pointer_pointer: *const *i32 = &pointer;
    const many_pointer: [*]i32 = @ptrCast(pointer);
    const null_pointer: ?*i32 = null;
    const alias_pointee: PointerAlias = 42;
    const alias_pointer: *const PointerAlias = &alias_pointee;
    const pair = PointerPair{ .first = 20, .second = 22 };
    const structure_pointer: *const PointerPair = &pair;
    const node = PointerNode{ .next = null, .value = 42 };
    const recursive_pointer: *const PointerNode = &node;
    const array = [2]i32{ 20, 22 };
    const array_pointer: *const [2]i32 = &array;
    const slice: []const i32 = &array;
    std.mem.doNotOptimizeAway(&pointer);
    std.mem.doNotOptimizeAway(&const_pointer);
    std.mem.doNotOptimizeAway(&pointer_pointer);
    std.mem.doNotOptimizeAway(&many_pointer);
    std.mem.doNotOptimizeAway(&null_pointer);
    std.mem.doNotOptimizeAway(&pointer_parameter);
    std.mem.doNotOptimizeAway(&alias_pointer);
    std.mem.doNotOptimizeAway(&structure_pointer);
    std.mem.doNotOptimizeAway(&recursive_pointer);
    std.mem.doNotOptimizeAway(&array_pointer);
    std.mem.doNotOptimizeAway(&slice);
    zig_sink = pointer_pointer.*.*;
    return pointer.* == 42 and pointer_parameter.* == 42 and
        pair.first + pair.second == 42 and node.next == null and node.value == 42 and
        array[0] + array[1] == 42;
}

pub fn main() u8 {
    const flag_ptr: *volatile bool = &input_flag;
    const signed_ptr: *volatile i32 = &input_signed;
    const unsigned_ptr: *volatile u64 = &input_unsigned;
    const single_ptr: *volatile f32 = &input_single;
    const double_ptr: *volatile f64 = &input_double;
    const succeeded = inspectScalars(
        flag_ptr.*,
        signed_ptr.*,
        unsigned_ptr.*,
        single_ptr.*,
        double_ptr.*,
    );
    const scoped = inspectScopes(signed_ptr.*);
    const pointers = inspectPointers(40, &pointer_parameter_value);
    return @intFromBool(!succeeded or !pointers or scoped != -41 or zig_constant != 37);
}

File: tests/fixtures/c/variables.c
#include <stdbool.h>

volatile int pointer_sink;
int pointer_parameter_value = 42;

typedef int aliased_int;

struct pointer_pair {
    int first;
    int second;
};

struct pointer_node {
    struct pointer_node *next;
    int value;
};

__attribute__((noinline)) static int pointer_identity(int value) {
    return value;
}

__attribute__((noinline)) static int parameter_target(int parameter) {
    int local = parameter + 1;
    return local;
}

__attribute__((noinline)) static int changing_target(void) {
    int changing = 10;
    for (int iteration = 0; iteration < 2; ++iteration) {
        changing += 7;
    }
    return changing;
}

__attribute__((noinline)) static int shadow_target(void) {
    int shadowed = 100;
    {
        int shadowed = 200;
        shadowed += 1;
        (void)shadowed;
    }
    return shadowed;
}

__attribute__((noinline)) static int partial_target(void) {
    int available = 42;
    int *pointer = &available;
    return *pointer;
}

__attribute__((noinline)) int pointer_target(int parameter, int *pointer_parameter) {
    int pointee = parameter + 2;
    int *pointer = &pointee;
    int **pointer_pointer = &pointer;
    const int *const_pointee = &pointee;
    int *const const_pointer = &pointee;
    void *void_pointer = &pointee;
    int *null_pointer = 0, *invalid_pointer = (int *)1;
    aliased_int alias_pointee = 42;
    aliased_int *alias_pointer = &alias_pointee;
    struct pointer_pair pair = {20, 22};
    struct pointer_pair *structure_pointer = &pair;
    struct pointer_node node = {0, 42};
    struct pointer_node *recursive_pointer = &node;
    int array[2] = {20, 22};
    int (*array_pointer)[2] = &array;
    int (*function_pointer)(int) = pointer_identity;
    __asm__ volatile("" : : "g"(pointer), "g"(pointer_pointer), "g"(const_pointee),
                     "g"(const_pointer), "g"(void_pointer), "g"(null_pointer),
                     "g"(invalid_pointer), "g"(pointer_parameter), "g"(alias_pointer),
                     "g"(structure_pointer), "g"(recursive_pointer), "g"(array_pointer),
                     "g"(function_pointer) : "memory");
    return **pointer_pointer + *pointer_parameter - 42;
}

__attribute__((noinline)) static int implicit_pointer_target(int parameter) {
    volatile int pointee = parameter + 2;
    volatile int *pointer = &pointee;
    volatile int **pointer_pointer = &pointer;
    pointer_sink = **pointer_pointer;
    return pointer_sink;
}

__attribute__((noinline)) static int implicit_pointer_offset_target(int parameter) {
    unsigned long storage = (unsigned long)(parameter + 2) << 32;
    unsigned char *byte_pointer = (unsigned char *)&storage + 4;
    pointer_sink = *byte_pointer;
    return pointer_sink;
}

int main(void) {
    _Bool boolean = true;
    char character = 65;
    signed char signed_character = -12;
    unsigned char unsigned_character = 250;
    short signed_short = -1234;
    unsigned short unsigned_short = 54321;
    int signed_int = -1234567;
    unsigned int unsigned_int = 3456789012U;
    long signed_long = -123456789L;
    unsigned long unsigned_long = 123456789UL;
    long long signed_long_long = -1234567890123LL;
    unsigned long long unsigned_long_long = 12345678901234ULL;
    float single = 1.25F;
    double double_precision = -2.5;
    long double extended = 3.125L;

    if (parameter_target(4) + changing_target() + shadow_target() + partial_target() +
            pointer_target(40, &pointer_parameter_value) +
            pointer_target(40, &pointer_parameter_value) + implicit_pointer_target(40) +
            implicit_pointer_offset_target(40) != 339) {
        return 1;
    }
    return boolean && character == 65 && signed_character == -12 &&
                   unsigned_character == 250 && signed_short == -1234 &&
                   unsigned_short == 54321 && signed_int == -1234567 &&
                   unsigned_int == 3456789012U && signed_long == -123456789L &&
                   unsigned_long == 123456789UL &&
                   signed_long_long == -1234567890123LL &&
                   unsigned_long_long == 12345678901234ULL && single == 1.25F &&
                   double_precision == -2.5 && extended == 3.125L
               ? 0
               : 1;
}
